Issue 105 web poet serialization in caching

lampssy opened the pull request:
Resolves #105 

I think this is more or less what we need to switch to a web-poet based caching. As for now, the tests are failing, as I didn't fully adjust them to the new cache. It would be good to get some insights whether the new implementation makes sense, before I proceed further.

wRAR commented on scrapy_poet/injection.py:
```diff
@@ -72,25 +73,22 @@ def load_providers(self, default_providers: Optional[Mapping] = None):  # noqa:
+
+        # SCRAPY_POET_CACHE: True
+        if cache_path and isinstance(cache_path, bool):
+            cache_path = get_scrapy_data_path(createdir=True)
```
I wonder if this should have some more specific name, like the old path including "scrapy-poet-cache" in the file name.

wRAR reviewed (commented):
Overall I think this approach is fine but I may be missing some special cases.

Gallaecio commented on scrapy_poet/cache.py:
```diff
@@ -20,54 +20,46 @@ def close(self) -> None:  # noqa: B027
 
-    def close(self) -> None:
-        self.db.close()
+    # TODO: Add option for compressed cache
```
@kmike @wRAR Do we want to implement this here or leave it for a future iteration?

wRAR replied:
I think it should be done separately and I see it as some kind of pluggable `SerializedDataFileStorage` implementations in web-poet.

Gallaecio reviewed (commented):
Amazing work so far!

wRAR commented:
> test_name_first_spider() fails because even though the cache is hit the spider gets the non-cached response

This is because test providers used by this test previously implemented the caching mixin and so returned their provided deps directly from the cache.

wRAR commented on scrapy_poet/page_input_providers.py:
```diff
@@ -133,6 +122,17 @@ def is_provided(self, type_: Callable) -> bool:
             )
 
+    @property
+    def has_cache_support(self):
```
This, and the code that uses it, assumes that if one dep is serializable then all of them are. Maybe we even don't need this function and should just check the deps directly when deciding whether to tryu caching them.

wRAR commented:
I've fixed the tests that were failing.

lampssy commented:
I added a more specific name for cache folder and removed checks whether providers have cache support - it looks like this is not needed.

@wRAR @Gallaecio let me know if you think any more adjustments needed.

Gallaecio reviewed (approved):
Seems good to me.

wRAR reviewed (approved):
Thanks!
